File: 20.valid_parentheses/valid.go
// 20. Valid Parentheses
// Topics: 'Stack', 'String'

// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

// An input string is valid if:

//     Open brackets must be closed by the same type of brackets.
//     Open brackets must be closed in the correct order.
//     Every close bracket has a corresponding open bracket of the same type.

// Example 1:

// Input: s = "()"

// Output: true

// Example 2:

// Input: s = "()[]{}"

// Output: true

// Example 3:

// Input: s = "(]"

// Output: false

// Example 4:

// Input: s = "([])"

// Output: true

// Example 5:

// Input: s = "([)]"

// Output: false

package validparentheses

import "github.com/volodimyr/leetcode/ds"

var ps = map[rune]rune{
	')': '(',
	']': '[',
	'}': '{',
}

func isValid(s string) bool {
	stack := &ds.Stack[rune]{}
	for _, r := range s {
		if r == '(' || r == '{' || r == '[' {
			stack.Push(r)
			continue
		}
		if ps[r] != stack.Pop() {
			return false
		}
	}
	return stack.Empty()
}

File: 768.valid_parenthesis_string/valid.py
# 678. Valid Parenthesis String
# Topics: 'String', 'Dynamic Programming', 'Stack', 'Greedy'
# Level: 'Medium'

# Given a string s containing only three types of characters: '(', ')' and '*', return true if s is valid.

# The following rules define a valid string:

#     Any left parenthesis '(' must have a corresponding right parenthesis ')'.
#     Any right parenthesis ')' must have a corresponding left parenthesis '('.
#     Left parenthesis '(' must go before the corresponding right parenthesis ')'.
#     '*' could be treated as a single right parenthesis ')' or a single left parenthesis '(' or an empty string "".

 

# Example 1:

# Input: s = "()"
# Output: true

# Example 2:

# Input: s = "(*)"
# Output: true

# Example 3:

# Input: s = "(*))"
# Output: true

 

# Constraints:

#     1 <= s.length <= 100
#     s[i] is '(', ')' or '*'.

class Solution:
    def checkValidString(self, s: str) -> bool:
        stack = []
        asterix = []

        for i in range(len(s)):
            ch = s[i]
            if ch == '(':
                stack.append(i)
            elif ch == '*':
                asterix.append(i)
            else:
                if stack:
                    stack.pop()
                elif not asterix:
                    return False
                else:
                    asterix.pop()
        
        while stack:
            if not asterix:
                return False
            if asterix[-1] < stack[-1]:
                return False
            stack.pop()
            asterix.pop()
        
        return True
        
